src/components/App/App.js

```diff
@@ -23,308 +23,336 @@ import newsApi from "../../utils/NewsApi";
+     setHasResults(true);
+    }
+   })
+   .catch((err) => {
+    console.log(err);
+    setHasError(true);
+   })
+   .finally(() => {
+    setIsLoading(false);
+   });
+ }
+
+ function handleLogOut() {
+  setLoggedIn(false);
+  localStorage.removeItem("jwt");
+  history.push("/");
+ }
+
+ function handleLogin() {
+  setHasError(false);
+  setLoggedIn(true);
+  setIsSignInOpen(false);
+ }
+
+ function handleRegister() {
+  setHasError(false);
+  setIsSignUpOpen(false);
+  setIsSuccessfulPopupOpen(true);
+ }
+
+ function handleSignInClick() {
+  setHasError(false);
+  setIsSignInOpen(true);
+  setIsSignUpOpen(false);
+  setIsSuccessfulPopupOpen(false);
+ }
+
+ function handleSignUpClick() {
+  setHasError(false);
+  setIsSignUpOpen(true);
+  setIsSignInOpen(false);
+ }
+
+ function handleRegisterSubmit(email, password, name) {
+  auth
+   .register(email, password, name)
+   .then((res) => {
+    if (res) {
+     setIsRegistered(true);
+     handleRegister();
+    } else {
+     setIsRegistered(false);
+     setHasError(true);
+    }
+   })
+   .catch((err) => {
+    console.log(`This email is unavailable: ${err.message}`);
+    setHasError(true);
+   });
+ }
+
+ function handleLoginSubmit(email, password) {
+  auth
+   .login(email, password)
+   .then((data) => {
+    if (data.token) {
+     localStorage.setItem("jwt", data.token);
+     setToken(data.token);
+     handleLogin();
+     history.push("/");
+    }
+   })
+   .catch((err) => {
+    console.log(`Incorrect email or password: ${err.message}`);
+    setHasError(true);
+   });
+ }
+
+ function closeAllPopups() {
+  setIsSignInOpen(false);
+  setIsSignUpOpen(false);
+  setIsSuccessfulPopupOpen(false);
+ }
+
+ return (
+  <CurrentUserContext.Provider value={currentUser}>
+   <div className="app">
+    <Header
+     loggedIn={loggedIn}
+     currentUser={currentUser}
+     setLoggedIn={setLoggedIn}
+     onSignInClick={handleSignInClick}
+     setIsNewsCardListOpen={setIsNewsCardListOpen}
+     setSearchKeyword={setSearchKeyword}
+     onSavedArticlesPage={onSavedArticlesPage}
+     onLogOut={handleLogOut}
+    />
+    <Switch>
+     <Route exact path="/">
+      <SearchForm
+       onSearch={handleSearchSubmit}
+       searchKeyword={searchKeyword}
+       setSearchKeyword={setSearchKeyword}
+       setIsNewsCardListOpen={setIsNewsCardListOpen}
+      />
+      {hasResults && isNewsCardListOpen && (
+       <NewsCardList
+        onSavedArticlesPage={onSavedArticlesPage}
+        loggedIn={loggedIn}
+        cards={cards}
+        savedArticles={savedArticles}
+        onSaveArticleClick={handleSaveArticle}
+        onRemoveArticleClick={handleRemoveArticle}
+        showCards={showCards}
+        setShowCards={setShowCards}
+        onSignInClick={handleSignInClick}
+       />
+      )}
+
+      {isLoading && <Preloader />}
+      {!hasResults && !isLoading && isNewsCardListOpen && (
+       <NotFoundResults hasError={hasError} />
+      )}
+      <About />
+     </Route>
+
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js
  function handleSubmit(request) {
    // start loading
    setIsLoading(true);
    request()
      // we need to close only in `then`
      .then(closeAllPopups)
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
      .catch(console.error)
      // and in finally we need to stop loading
      .finally(() => setIsLoading(false));
  }

```


You need to pass the **request** into the call of `handleSubmit` like that:

```js
 // here is an example
 function handleProfileFormSubmit(inputValues) {
  // here we create a function that returns a promise 
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally`
    return api.editProfile(inputValues).then(setCurrentUser);
  }
  // here we call handleSubmit passing the request
  handleSubmit(makeRequest);
}


```

So, this way you can remove a lot of code duplicating